server: TestStatusAPIStatements failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/93d3746f6f503f0896779514094338ea4f62739c

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStatusAPIStatements PKG=github.com/cockroachdb/cockroach/pkg/server TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1244121&tab=buildLog

```
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:631 +0x53
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:639 +0x34
==================
I190416 07:56:05.345195 45551 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] transport racer
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.346454 45552 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
W190416 07:56:05.348206 44401 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
W190416 07:56:05.348780 44690 gossip/gossip.go:1496  [n3] no incoming or outgoing connections
W190416 07:56:05.349800 44941 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 2: rpc error: code = Canceled desc = context canceled:
W190416 07:56:05.350814 45135 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = context canceled:
W190416 07:56:05.349895 45495 storage/intentresolver/intent_resolver.go:822  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/24/RangeDescriptor: node unavailable; try another peer
W190416 07:56:05.353471 45067 storage/raft_transport.go:583  [n2] while processing outgoing Raft queue to node 1: EOF:
I190416 07:56:05.345246 45553 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.355738 45551 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
W190416 07:56:05.356824 45188 storage/raft_transport.go:583  [n3] while processing outgoing Raft queue to node 1: EOF:
I190416 07:56:05.360874 44230 rpc/nodedialer/nodedialer.go:143  [n1] unable to connect to n2: failed to connect to n2 at 127.0.0.1:38475: context canceled
I190416 07:56:05.361924 44230 rpc/nodedialer/nodedialer.go:143  [n1] unable to connect to n3: failed to connect to n3 at 127.0.0.1:39627: context canceled
I190416 07:56:05.362434 45553 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.362535 45551 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.362849 45551 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.363583 45551 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] transport racer
I190416 07:56:05.367444 45553 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.367766 45553 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-subscription
I190416 07:56:05.367931 45552 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.375199 45552 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I190416 07:56:05.447118 44094 kv/transport_race.go:113  transport race promotion: ran 51 iterations on up to 877 requests
--- FAIL: TestStatusAPIStatements (6.05s)
    testing.go:771: race detected during execution of test

```

--- a/pkg/sql/app_stats.go
+++ b/pkg/sql/app_stats.go
@@ -351,11 +351,11 @@ func (s *sqlStats) getStmtStats(
 	var ret []roachpb.CollectedStatementStatistics
 	salt := ClusterSecret.Get(&s.st.SV)
 	for appName, a := range s.apps {
+		a.Lock()
 		if cap(ret) == 0 {
 			// guesstimate that we'll need apps*(queries-per-app).
 			ret = make([]roachpb.CollectedStatementStatistics, 0, len(a.stmts)*len(s.apps))
 		}
-		a.Lock()
 		for q, stats := range a.stmts {
 			maybeScrubbed := q.stmt
 			maybeHashedAppName := appName

